Allow prescribed velocities with locally conservative discretization.

maxrudolph opened the pull request:
This implements the fix to the prescribed internal velocities plugin as suggested by @bangerth , in response to #1876 .

bangerth commented on cookbooks/prescribed_velocity/corner_flow.prm:
```diff
@@ -5,6 +5,11 @@ set End time    = 0
 # Load the signal library.
 set Additional shared libraries = ./libprescribed_velocity.so
 
+subsection Discretization
```
Instead of modifying an existing test, can you just clone it? This way, we test both the existing code path as well as the new one of this patch.

maxrudolph replied:
This is just the cookbook input file, not a test. I am happy to add a test for the DGP case if that would be useful.

bangerth replied:
Ah, I see. But why change the cookbook?

As for test: Yes, that would be very useful indeed -- we usually try to exercise all code paths in the testsuite somehow.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
   }
 
   /**
+   * This function retrieves the unit support points (in the unit cell) for the current element
```
period at the end of the sentence.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
 
   /**
+   * This function retrieves the unit support points (in the unit cell) for the current element
+   * The DGP elements used when 'set Use locally conservative discretization = true' is enabled do
```
"is enabled do" should probably be "is enabled to", but that still sounds funny. How about "...does not have support points"

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
   /**
+   * This function retrieves the unit support points (in the unit cell) for the current element
+   * The DGP elements used when 'set Use locally conservative discretization = true' is enabled do
+   * not have support points. If these elements are in use, a ficticious support point at the cell
```
-> fictitious

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
+   * This function retrieves the unit support points (in the unit cell) for the current element
+   * The DGP elements used when 'set Use locally conservative discretization = true' is enabled do
+   * not have support points. If these elements are in use, a ficticious support point at the cell
+   * center is returned. This is necessary for the internal velocities to be constrained as desired.
```
-> is returned for each shape function that corresponds to the pressure variable, whereas the support points for the velocity are correct; the fictitious points don't matter because we only use this function when interpolating the velocity variable, and ignore the evaluation at the pressure support points.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
+    std::vector< Point<dim> > unit_support_points;
+    if ( !use_locally_conservative_discretization )
+      {
+        unit_support_points = simulator_access.get_fe().get_unit_support_points();
```
or, more easily,
```
  return simulator_access.get_fe().get_unit_support_points();
```
Then you can move the declaration of the `unit_support_points` variable into the `else` block, along with the `return` statement at the bottom of the function.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
+      {
+        //special case for discontinuous pressure elements, which lack unit support points
+        const unsigned int dofs_per_cell = simulator_access.get_fe().dofs_per_cell;
+        for (unsigned int dof=0; dof < dofs_per_cell; dof++)
```
for consistency with the rest of the code base, use `++dof`.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -110,6 +110,51 @@ namespace aspect
+                if ( dim==2 )
+                  unit_support_points.push_back( Point<dim> (0.5,0.5) );
+                if ( dim==3 )
+                  unit_support_points.push_back( Point<dim> (0.5,0.5,0.5) );
```
on second thought, rather than making the distinction between dimensions, just do
```
  // manufacture a support point, arbitrarily at cell's origin
  unit_support_points.push_back (Point<dim>());
```
This works independent of the dimension and is shorter.

maxrudolph replied:
Will the default constructor for point place the point at 0.5,0.5 or at some other value like null? There is actually a reason why I wrote the code this way. I want to place the manufactured point at the cell center so that in my `prescribed_velocity_pressure.cc` plugin ( #1864 ) which reuses this code, the pressure indicator function is evaluated at a meaningful location.

bangerth replied:
`Point<dim>()` creates (0,0) or (0,0,0). If you like the center of the cell better, then leave it as is.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -120,7 +165,11 @@ namespace aspect
+        const Parameters<dim> parameters = simulator_access.get_parameters();
+        const bool use_locally_conservative_discretization = parameters.use_locally_conservative_discretization;
+
+        std::vector< Point<dim> > points = get_unit_support_points_for_velocity<dim>(simulator_access,use_locally_conservative_discretization);
```
* no need to explicitly say `<dim>` in the function call.

* mark the variable as `const`

* also, can you break at the comma between function arguments? The line is long enough.

bangerth commented on cookbooks/prescribed_velocity/prescribed_velocity.cc:
```diff
@@ -120,7 +165,11 @@ namespace aspect
+        const bool use_locally_conservative_discretization = parameters.use_locally_conservative_discretization;
+
+        std::vector< Point<dim> > points = get_unit_support_points_for_velocity<dim>(simulator_access,use_locally_conservative_discretization);
+        Quadrature<dim> quadrature (points);
```
while you're there, make this `const` as well.

maxrudolph commented:
@bangerth This commit addresses your comments above and adds a test for the DGP case. I also updated the test for the default element to use the new prescribed velocity plugin code.

maxrudolph commented:
@bangerth I think that this is ready for a final review and merge.

gassmoeller commented:
/run-tests

gassmoeller commented:
Makes total sense. Ready to merge when the tester is happy. Sorry for coming back so late to this PR.
